File: lib/src/native/qa_media_span.dart
import 'dart:ffi';
import 'dart:typed_data';

import 'package:ffi/ffi.dart';
import 'package:flutter/foundation.dart' show visibleForTesting;

import 'native_scratch.dart';
import 'qa_engine_abi.dart';

/// A medium stored in a span of a file — its own bytes, or FRAMED in
/// compressed blocks — read a window at a time by the engine's one reader of
/// both (`qa_media_span.c`).
///
/// 🚨★★★**THE READER THE DECODERS USE, AND NOW DART'S TOO.** A framed medium
/// was read by a Dart copy of the block walk for everything Dart reads (a
/// PDF's pages, a picture, a fingerprint) and by nothing at all on the
/// native side, which is why a carried movie kept compressed could not be
/// played and a framed sound was assembled whole before it was decoded
/// (board `carried-movie-compressed`, 유저 2026-09-24 「최대한 통일할거
/// 통일해줘」). Both sides read through this one now, so a Dart read and a
/// decoder's read of one span cannot come to disagree.
///
/// ⛔A PLAIN span is read through here only by native code. Dart reads a
/// plain file with `RandomAccessFile` (`MediaByteSource`), because a build
/// without the engine still has to read one; a framed medium has no such
/// floor — its blocks are zstd, which only the engine carries.
final class QaMediaSpan {
  QaMediaSpan._(this._api, this._handle, this.size);

  final _QaMediaSpanApi _api;
  Pointer<Void> _handle;

  /// How many bytes the MEDIUM has — for a framed span, decoded.
  final int size;

  /// Whether this build can read a span at all.
  static bool get available => _QaMediaSpanApi.instance != null;

  /// The medium stored in `[offset, offset + length)` of [path] — its own
  /// bytes, or a framed blob of them when [framed]. Null when there is no
  /// engine, or when the span does not hold together: not inside the file,
  /// or a framed header that disagrees with the span it sits in.
  static QaMediaSpan? open(
    String path, {
    required int offset,
    required int length,
    required bool framed,
  }) {
    final api = _QaMediaSpanApi.instance;
    if (api == null) {
      return null;
    }
    final utf8Path = path.toNativeUtf8(allocator: malloc);
    final Pointer<Void> handle;
    try {
      handle = api.open(utf8Path, offset, length, framed ? 1 : 0);
    } finally {
      malloc.free(utf8Path);
    }
    if (handle == nullptr) {
      return null;
    }
    return QaMediaSpan._(api, handle, api.size(handle));
  }

  /// Up to [count] bytes of the medium from [position] into [buffer]: how
  /// many landed, short only at the end and 0 past it. Throws a
  /// [FormatException] when the file would not read or a block would not
  /// decode — a broken medium says so rather than handing back less.
  ///
  /// ⚠️Through a scratch of at most [_windowBytes] at a time, so reading a
  /// whole medium does not also park one of its size in native memory.
  int readInto(Uint8List buffer, int position, int count) {
    if (_handle == nullptr) {
      throw StateError('read from a closed media span');
    }
    var wrote = 0;
    while (wrote < count) {
      final want = count - wrote < _windowBytes ? count - wrote : _windowBytes;
      final scratch = _window.ensure(want);
      final got = _api.read(_handle, position + wrote, scratch.cast(), want);
      if (got < 0) {
        throw const FormatException('the stored medium would not read back');
      }
      if (got == 0) {
        break;
      }
      buffer.setRange(wrote, wrote + got, scratch.asTypedList(got));
      wrote += got;
      if (got < want) {
        break;
      }
    }
    return wrote;
  }

  /// What this span has cost so far: blocks decoded, and stored bytes read
  /// (the framed header included) — what tells a reader that reads only the
  /// blocks a window lands in from one that quietly reads them all.
  ({int blocksDecoded, int storedBytesRead}) get stats {
    final decoded = calloc<Int64>();
    final read = calloc<Int64>();
    try {
      _api.stats(_handle, decoded, read);
      return (blocksDecoded: decoded.value, storedBytesRead: read.value);
    } finally {
      calloc
        ..free(decoded)
        ..free(read);
    }
  }

  void close() {
    if (_handle != nullptr) {
      debugOnClose?.call(stats);
      _api.close(_handle);
      _handle = nullptr;
    }
  }

  /// Test hook: handed what each span cost as it closes — the one way to
  /// count what a ONE-SHOT read spent, since the span it read through is
  /// gone by the time the read returns.
  @visibleForTesting
  static void Function(({int blocksDecoded, int storedBytesRead}) cost)?
  debugOnClose;

  /// The most one native read hands over at a time.
  static const int _windowBytes = 1024 * 1024;

  /// ⚠️One for the isolate, not one per span: nothing here suspends between
  /// the native read and the copy out, and a scratch per span would be one
  /// more buffer per open PDF.
  static final NativeScratch<Uint8> _window = NativeScratch<Uint8>(
    (count) => calloc<Uint8>(count),
  );
}

/// The span reader's exports, resolved once per isolate.
final class _QaMediaSpanApi {
  _QaMediaSpanApi._(DynamicLibrary library)
    : open = library
          .lookupFunction<
            Pointer<Void> Function(Pointer<Utf8>, Int64, Int64, Int32),
            Pointer<Void> Function(Pointer<Utf8>, int, int, int)
          >('qa_media_span_open'),
      size = library
          .lookupFunction<
            Int64 Function(Pointer<Void>),
            int Function(Pointer<Void>)
          >('qa_media_span_size'),
      read = library
          .lookupFunction<
            Int64 Function(Pointer<Void>, Int64, Pointer<Void>, Int64),
            int Function(Pointer<Void>, int, Pointer<Void>, int)
          >('qa_media_span_read'),
      close = library
          .lookupFunction<
            Void Function(Pointer<Void>),
            void Function(Pointer<Void>)
          >('qa_media_span_close'),
      stats = library
          .lookupFunction<
            Void Function(Pointer<Void>, Pointer<Int64>, Pointer<Int64>),
            void Function(Pointer<Void>, Pointer<Int64>, Pointer<Int64>)
          >('qa_media_span_stats');

  final Pointer<Void> Function(Pointer<Utf8>, int, int, int) open;
  final int Function(Pointer<Void>) size;
  final int Function(Pointer<Void>, int, Pointer<Void>, int) read;
  final void Function(Pointer<Void>) close;
  final void Function(Pointer<Void>, Pointer<Int64>, Pointer<Int64>) stats;

  static _QaMediaSpanApi? _instance;
  static bool _tried = false;

  static _QaMediaSpanApi? get instance {
    if (!_tried) {
      _tried = true;
      final library = openQaEngineLibrary();
      try {
        _instance = library == null ? null : _QaMediaSpanApi._(library);
      } on ArgumentError {
        _instance = null;
      }
    }
    return _instance;
  }
}

File: packages/qa_native/src/qa_media_span.c
// The one reader of a medium stored in a span of a file — see
// qa_media_span.h for why there is only one, and for the framed layout.

#include "qa_media_span.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "qa_platform_path.h"
#include "third_party/zstd/zstd.h"

/// Bytes before the block lengths: blockBytes, totalLength, blockCount.
#define QA_FRAMED_PREFIX 16

/// A block size no writer of this format has used, and no reader should
/// trust — the header is read from a file, and a number this size would
/// have one allocation take everything.
#define QA_FRAMED_MAX_BLOCK (64 * 1024 * 1024)

struct qa_media_span {
  FILE* file;
  /// Where the STORED bytes start in the file, and how many there are.
  int64_t base;
  int64_t stored_length;
  /// How many bytes the MEDIUM has.
  int64_t size;
  int32_t framed;

  // ---- framed only
  int64_t block_bytes;
  int64_t block_count;
  /// block_count + 1 offsets into the stored bytes: where each block starts,
  /// and where the last one ends.
  int64_t* block_starts;
  /// One compressed block, sized to the largest.
  uint8_t* stored_block;
  /// One decoded block — the one the last window landed in, kept for the
  /// next: a document reads in windows far smaller than a block, and
  /// decoding 512KB for each of them is the cost this exists to not pay.
  uint8_t* block;
  int64_t block_index;
  int64_t block_size;

  int64_t blocks_decoded;
  int64_t stored_bytes_read;
};

static uint32_t qa_read_u32(const uint8_t* at) {
  return (uint32_t)at[0] | ((uint32_t)at[1] << 8) | ((uint32_t)at[2] << 16) |
         ((uint32_t)at[3] << 24);
}

static uint64_t qa_read_u64(const uint8_t* at) {
  return (uint64_t)qa_read_u32(at) | ((uint64_t)qa_read_u32(at + 4) << 32);
}

/// Exactly [want] stored bytes from [at] (relative to the span's base), or 0.
static int qa_span_read_stored(qa_media_span* span,
                               int64_t at,
                               void* out,
                               int64_t want) {
  if (at < 0 || want < 0 || at + want > span->stored_length) {
    return 0;
  }
  if (!qa_seek_absolute(span->file, span->base + at)) {
    return 0;
  }
  int64_t got = 0;
  while (got < want) {
    const size_t read =
        fread((uint8_t*)out + got, 1, (size_t)(want - got), span->file);
    if (read == 0) {
      return 0;
    }
    got += (int64_t)read;
  }
  span->stored_bytes_read += want;
  return 1;
}

/// Reads the framed header and sizes the buffers, or answers 0 when it does
/// not hold together with the span it sits in.
static int qa_span_open_framed(qa_media_span* span) {
  uint8_t prefix[QA_FRAMED_PREFIX];
  if (!qa_span_read_stored(span, 0, prefix, QA_FRAMED_PREFIX)) {
    return 0;
  }
  const int64_t block_bytes = (int64_t)qa_read_u32(prefix);
  const uint64_t total = qa_read_u64(prefix + 4);
  const int64_t block_count = (int64_t)qa_read_u32(prefix + 12);
  if (block_bytes <= 0 || block_bytes > QA_FRAMED_MAX_BLOCK ||
      total > (uint64_t)INT64_MAX) {
    return 0;
  }
  // ⛔The count must be the one the length implies: a header that says
  // otherwise was not written by this format's writer.
  const int64_t implied = ((int64_t)total + block_bytes - 1) / block_bytes;
  if (block_count != implied) {
    return 0;
  }
  const int64_t header_length = QA_FRAMED_PREFIX + 4 * block_count;
  if (header_length > span->stored_length) {
    return 0;
  }
  uint8_t* lengths = NULL;
  if (block_count > 0) {
    lengths = (uint8_t*)malloc((size_t)(4 * block_count));
    if (lengths == NULL ||
        !qa_span_read_stored(span, QA_FRAMED_PREFIX, lengths,
                             4 * block_count)) {
      free(lengths);
      return 0;
    }
  }
  span->block_starts =
      (int64_t*)malloc((size_t)(block_count + 1) * sizeof(int64_t));
  if (span->block_starts == NULL) {
    free(lengths);
    return 0;
  }
  int64_t at = header_length;
  int64_t largest = 0;
  for (int64_t i = 0; i < block_count; i += 1) {
    const int64_t length = (int64_t)qa_read_u32(lengths + 4 * i);
    span->block_starts[i] = at;
    at += length;
    if (length <= 0 || at > span->stored_length) {
      // A block that runs past the span is a torn entry, and reading it
      // would decode whatever follows it in the file.
      free(lengths);
      return 0;
    }
    if (length > largest) {
      largest = length;
    }
  }
  span->block_starts[block_count] = at;
  free(lengths);
  span->block_bytes = block_bytes;
  span->block_count = block_count;
  span->size = (int64_t)total;
  if (block_count > 0) {
    span->stored_block = (uint8_t*)malloc((size_t)largest);
    span->block = (uint8_t*)malloc((size_t)block_bytes);
    if (span->stored_block == NULL || span->block == NULL) {
      return 0;
    }
  }
  span->block_index = -1;
  return 1;
}

qa_media_span* qa_media_span_open(const char* utf8_path,
                                  int64_t offset,
                                  int64_t length,
                                  int32_t framed) {
  if (utf8_path == NULL || offset < 0 || length < 0) {
    return NULL;
  }
  FILE* file = qa_open_path_read(utf8_path);
  if (file == NULL) {
    return NULL;
  }
  // ⛔The span must be INSIDE the file. One that promises bytes the file
  // cannot supply is a truncated medium, which reads as a corrupt one rather
  // than as a span that was wrong.
  const int64_t file_size = qa_file_size(file);
  if (file_size < 0 || offset + length > file_size) {
    fclose(file);
    return NULL;
  }
  qa_media_span* span = (qa_media_span*)calloc(1, sizeof(*span));
  if (span == NULL) {
    fclose(file);
    return NULL;
  }
  span->file = file;
  span->base = offset;
  span->stored_length = length;
  span->size = length;
  span->framed = framed ? 1 : 0;
  span->block_index = -1;
  if (span->framed && !qa_span_open_framed(span)) {
    qa_media_span_close(span);
    return NULL;
  }
  return span;
}

int64_t qa_media_span_size(const qa_media_span* span) {
  return span == NULL ? 0 : span->size;
}

/// The decoded block [index], through the one-block cache, or NULL.
static const uint8_t* qa_span_block(qa_media_span* span, int64_t index) {
  if (span->block_index == index) {
    return span->block;
  }
  const int64_t start = span->block_starts[index];
  const int64_t stored = span->block_starts[index + 1] - start;
  // The last block is short; every other one is exactly block_bytes.
  const int64_t expected = index == span->block_count - 1
                               ? span->size - span->block_bytes * index
                               : span->block_bytes;
  span->block_index = -1;
  if (!qa_span_read_stored(span, start, span->stored_block, stored)) {
    return NULL;
  }
  const size_t decoded =
      ZSTD_decompress(span->block, (size_t)span->block_bytes,
                      span->stored_block, (size_t)stored);
  // ⛔A block that decodes to any other length is a broken file, not a
  // short one: the window it serves would be the wrong bytes.
  if (ZSTD_isError(decoded) || (int64_t)decoded != expected) {
    return NULL;
  }
  span->blocks_decoded += 1;
  span->block_index = index;
  span->block_size = expected;
  return span->block;
}

int64_t qa_media_span_read(qa_media_span* span,
                           int64_t position,
                           void* out,
                           int64_t want) {
  if (span == NULL || out == NULL || position < 0 || want <= 0 ||
      position >= span->size) {
    return 0;
  }
  if (want > span->size - position) {
    want = span->size - position;
  }
  if (!span->framed) {
    // 🚨base + position. Reading at `position` alone hands back whatever
    // precedes the medium in the file as if it were the medium.
    return qa_span_read_stored(span, position, out, want) ? want : -1;
  }
  int64_t wrote = 0;
  while (wrote < want) {
    const int64_t at = position + wrote;
    const int64_t index = at / span->block_bytes;
    const uint8_t* block = qa_span_block(span, index);
    if (block == NULL) {
      return -1;
    }
    const int64_t from = at - index * span->block_bytes;
    int64_t take = span->block_size - from;
    if (take > want - wrote) {
      take = want - wrote;
    }
    memcpy((uint8_t*)out + wrote, block + from, (size_t)take);
    wrote += take;
  }
  return wrote;
}

void qa_media_span_close(qa_media_span* span) {
  if (span == NULL) {
    return;
  }
  if (span->file != NULL) {
    fclose(span->file);
  }
  free(span->block_starts);
  free(span->stored_block);
  free(span->block);
  free(span);
}

void qa_media_span_stats(const qa_media_span* span,
                         int64_t* blocks_decoded,
                         int64_t* stored_bytes_read) {
  if (blocks_decoded != NULL) {
    *blocks_decoded = span == NULL ? 0 : span->blocks_decoded;
  }
  if (stored_bytes_read != NULL) {
    *stored_bytes_read = span == NULL ? 0 : span->stored_bytes_read;
  }
}
